cpu/sam0_common/periph/i2c.c

```diff
@@ -471,46 +315,51 @@ static inline int _write(SercomI2cm *dev, const uint8_t *data, int length)
 
         /* Wait for response on bus. */
         if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -1;
+            return I2C_ERR;
 
         /* Check for NACK from slave. */
         if (dev->STATUS.reg & SERCOM_I2CM_STATUS_RXNACK) {
             DEBUG("STATUS_ERR_OVERFLOW\n");
-            return -4;
+            return I2C_DATA_NACK;
         }
     }
-    return 0;
+    return I2C_ACK;
 }
 
-static inline int _read(SercomI2cm *dev, uint8_t *data, int length)
+static inline int _read(SercomI2cm *dev, uint8_t *data, int length, uint8_t stop)
 {
     uint8_t count = 0;
 
     /* Set action to ack. */
     dev->CTRLB.reg &= ~SERCOM_I2CM_CTRLB_ACKACT;
 
     /* Read data buffer. */
-    while (count != length) {
+    while (length--) {
         /* Check that bus ownership is not lost. */
         if ((dev->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) != BUSSTATE_OWNER) {
             DEBUG("STATUS_ERR_PACKET_COLLISION\n");
-            return -2;
+            return I2C_ERR;
         }
 
         /* Wait for hardware module to sync */
         while (dev->SYNCBUSY.reg & SERCOM_I2CM_SYNCBUSY_MASK) {}
+        /* Check if this is the last byte to read */
+        if (length == 0 && stop) {
+            /* Send NACK before STOP */
+            dev->CTRLB.reg |= SERCOM_I2CM_CTRLB_ACKACT;
+            /* Prepare stop command before read last byte */
+            _stop(dev);
+        }
         /* Save data to buffer. */
         data[count] = dev->DATA.reg;
 
         /* Wait for response on bus. */
         if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -1;
+            return I2C_ERR;
```

Comment:
It seems to always fail here, maybe put a condition that it only checks if it is not the last byte.